### ISSUE TITLE
Relinquish leadership based on conditions

### ISSUE THREAD

**Issue by jyotimahapatra** [kubernetes/kubernetes#111899]
### What happened?

A kubernetes cluster deployed in AWS is subject to occasional zonal [outages](https://aws.amazon.com/premiumsupport/technology/pes/). If the kubernetes control plane(master) instances are part of the zone experiencing outage, the cluster can have downtime with behaviors defined by the specific component that is failing during the outage. In the most common cluster setup usecases(e.g kops), all control plane components(e.g kube-apiserver, kube-scheduler, kube-controller-manager) are setup as static pods. Depending on how the cluster is setup further, there could be more components(e.g webhooks, authenticator, cloud-controller-manager, etc) running as static pods too. 

During zonal outages that do not affect apiserver to etcd connectivity, any controller that depends on leader election and is currently hosted on an instance in the affected zone continues to renew its lease. Continuing to hold the lease is fine as long as the controller's workload is only dependent on apiserver objects. In cases where the controller's work depends on aws api calls, success of the controller is not guaranteed. In such cases the workloads would benefit from lease transfer.

As an operator of clusters in AWS, it might be beneficial if there was a way to indicate the controllers using abstract rules (that work across cloud) to exclude controllers from taking lease if certain rules match. The rule can be created as a separate api object that cluster operator can control. During zonal outages, the operator can refer to the aws status page and indicate any lease based controllers to evacuate from the bad zone. 

Let me your thoughts on the idea.

/sig-apimachinary
 

### What did you expect to happen?

-

### How can we reproduce it (as minimally and precisely as possible)?

-

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```

</details>

### Cloud provider

<details>
AWS
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by leilajal** [comment]
/cc @roycaihw 
during triage we were thinking you might have an opinion here. thanks!
/triage accepted

**Comment by deads2k** [comment]
If a given process cannot fulfill its responsibilities, I would expect that process to go livez=false and get shutdown.  When it is shutdown, it will release its lease and another instance can acquire the lease.

**Comment by jyotimahapatra** [comment]
some more context about the problem:

kubernetes clusters running in the cloud typically are subjected to errors in failure domains. When zonal outages occur, they affect the running workloads in the affected zone in various forms. The effect depends on the kind of dependencies of the pods and the impact of the zonal outage on the dependencies. For pods that serve live traffic, cluster operators can use many mechanisms to shift traffic from the affected zone. However, there is no way to make controllers that depend on leader election to be weighed away. Over many zonal outages in AWS over the past year, there were a few learnings. If a zonal outage causes node termination or network partition such that the leader pod cannot connect with the api-server, the leadership naturally shifts away. This works for most use cases. However, when the zonal outage does not cause network connectivity problems between the pod and apiserver, the lease object is held by a bad pod, even if it cannot perform its duties. Some examples from recent observations in AWS zonal outage in eu-west-2 and us-east-2. 
A custom scheduler backed by leadership held onto the lease, but was not able to make aws api calls. There is no heuristic defined that would cause the health check to fail for such a custom controller. For the duration of the outage, the scheduler was unable to perform any action.
kube-controller-manager ‘s node controller maintained a watch on the kube-apiserver. During the outage the etcd had a network partition and had a split brain. The watch persisted on the partitioned etcd node. The partitioned etcd node of course did not have any further updates and KCM’s node controller did not receive any further updates. After the network partition recovered, the watch caught up. A livez check on KCM could not have identified this issue.

Although these two would look like two isolated events, the pattern is that no controller is self sufficient to know all the error cases. As an operator of a many clusters, it would have been great if we have a mechanism to tell lease controller that a zone is impacted. As an operator I don;t have the ability to modify or guarantee that the liveness check can capture all the nuanced error cases that can happen during such outages. 

Taking inspiration from [scheduler topology spread constraints](https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/) we could introduce an object that leader election mechanism can use to detect evicted topologies. This would be an opt in mechanism and work safely only if all pods that are candidates for a leader are in different zones. I was able to also [simulate](https://github.com/kubernetes/kubernetes/compare/v1.22.9...jyotimahapatra:kubernetes:22.9-leader?expand=1) the proposed behavior using labels to guide eviction based on topology.

**Comment by jyotimahapatra** [comment]
@deads2k Following up on the suggestions from api-machinery meeting on 08/24, i explored lease stealing technique.
Today the id for lock holder is based on `os.HostName`([link](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L242)). Would we be open to allow changing the id to some other configurable mechanism(flag?) so that an operator can use an well known pattern to know when to steal?

**Comment by dims** [comment]
cc @dims

**Comment by ellistarn** [comment]
> If a given process cannot fulfill its responsibilities, I would expect that process to go livez=false and get shutdown. When it is shutdown, it will release its lease and another instance can acquire the lease.

Doesn't this create a race condition between the lease acquisition and the liveness probe? What if the kubelet's probe is delayed? While liveness probes are one of the primary mechanisms to force a process to terminate (without deleting the pod; we don't want hot failover), it's a fairly blunt instrument to achieve our goal.

We have dozens of controllers that need to relinquish leadership during an AZ gray failure (related: https://github.com/aws/karpenter/issues/4727). Ideally, we could have a common solution. One we have brainstormed so far:

1. AZ detection system could annotate a lease `kubernetes.io/reject-holder-identity=current-holder`

**Comment by cici37** [comment]
Hi @jyotimahapatra , thanks for the issue. Are you still interested in moving forward? If so, discussion in the sig meeting sounds like a reasonable next step :)
/triage accepted

**Comment by jyotimahapatra** [comment]
yes. The intention of this issue as mentioned in the initial post `if there was a way to indicate the controllers using abstract rules (that work across cloud) to exclude controllers from taking lease if certain rules match` is coming to fruition in a different way [here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/coordination/v1/types.go#L80-L86). If the impl of coordinated leases is extensible to accomodate more predicates , then the needed usecase will be solved. 

I'll evaluate the coordinated lease alpha API and report back about my observations.
